cmi.proto

```diff
@@ -7,236 +7,273 @@ import "google/protobuf/wrappers.proto";
+message ProbeRequest {
+    // Intentionally empty.
 }
 
-message ListMachinesResponse {
-  // MachineList is the map of list of machines. Format for the map should be map<MachineID, MachineName>.
-  // This field is REQUIRED.
-  map<string, string> MachineList = 1;
+message ProbeResponse {
+    // Readiness allows a plugin to report its initialization status back
+    // to the CMI-Client. Initialization for some plugins MAY be time consuming
+    // and it is important for a CMI-Client to distinguish between the following
+    // cases:
+    //
+    // 1) The plugin is in an unhealthy state and MAY need restarting. In
+    //    this case a gRPC error code SHALL be returned.
+    // 2) The plugin is still initializing, but is otherwise perfectly
+    //    healthy. In this case a successful response SHALL be returned
+    //    with a readiness value of `false`. Calls to the plugin's
+    //    Controller and/or Node services MAY fail due to an incomplete
+    //    initialization state.
+    // 3) The plugin has finished initializing and is ready to service
+    //    calls to its Controller and/or Node services. A successful
+    //    response is returned with a readiness value of `true`.
+    //
+    // This field is OPTIONAL. If not present, the caller SHALL assume
+    // that the plugin is in a ready state and is accepting calls to its
+    // Controller and/or Node services (according to the plugin's reported
+    // capabilities).
+    .google.protobuf.BoolValue ready = 1;
 }
 
-message ShutDownMachineRequest {
-  // MachineID is the id of the machine to be shut-down.
-  // It should uniquely identify the real machine in cloud-provider.
-  // This field is REQUIRED.
-  string MachineID = 1;
-
-  // Secrets is the map containing necessary credentials for cloud-provider to delete the machine.
-  // This field is OPTIONAL.
-  map<string, bytes> Secrets = 2 [(cmi_secret) = true];
-}
+//---------  Machine Service -------//
 
-message ShutDownMachineResponse {
-  // Intentionally empty.
+service Machine {
+    // CreateMachine is the gRPC call to create a machine.
+    // This method is REQUIRED.
+    rpc CreateMachine (CreateMachineRequest)
+        returns (CreateMachineResponse) {}
+
+    // DeleteMachine is the gRPC call to delete a machine.
+    // This method is REQUIRED.
+    rpc DeleteMachine (DeleteMachineRequest)
+        returns (DeleteMachineResponse) {}
+
+    // GetMachine is the gRPC call to get the status of a machine.
+    // This method is REQUIRED.
+    rpc GetMachine (GetMachineRequest)
+        returns (GetMachineResponse) {}
+
+    // ListMachines is the gRPC call to list machines backed by the providerSpec.
+    // This method is REQUIRED.
+    rpc ListMachines (ListMachinesRequest)
+        returns (ListMachinesResponse) {}
+
+    // GetListOfVolumeIDsForExistingPVs() is the gRPC call to list VolumeIDs
+    // (if exists) for PVs provided by PVSpecList.
+    // This method is REQUIRED.
```

Comment:
Are we planning to make this `REQUIRED` or `OPTIONAL`?

Reply:
I am not sure. What's your opinion?

Reply:
If `GetListOfVolumeIDsForExistingPVs()` is not implemented, then the drain logic will NOT wait for PV to detach while draining a node. If this is implemented, then the drain logic WILL ALWAYS wait for PV to detach while machine drain. There is no other effect of this method as of now.
Any of the above scenarios could be desired depending on the cloud provider (or even workload).
IMHO, this method should be optional.

/cc @amshuman-kr

Reply:
:+1: Agreed. Also, we can take onboard @stoyanr's suggestion to rename the name of the this method to be more succinct and clear.